File: pkg/lattice/types/asset.go
// Asset is a dataset declaration: schema, quality rules, ingestion sink, and
// optional column mapping. Assets are populated by the upload pipeline.

package types

import (
	"encoding/json"
	"time"
)

// AssetID is the canonical handle for an asset.
type AssetID string

// Asset is the metadata that drives the upload + ingestion pipeline.
type Asset struct {
	ID                   AssetID           `json:"id"`
	WorkspaceID          WorkspaceID       `json:"workspace_id"`
	APIName              APIName           `json:"api_name"`
	DisplayName          string            `json:"display_name,omitempty"`
	Description          string            `json:"description,omitempty"`
	Metadata             map[string]any    `json:"metadata,omitempty"`
	Tags                 []string          `json:"tags,omitempty"`
	Properties           []Property        `json:"properties"`
	QualityRules         []QualityRule     `json:"quality_rules,omitempty"`
	Dependencies         []AssetDependency `json:"dependencies,omitempty"`
	Sink                 AssetSink         `json:"sink"`
	SavedColumnMapping   []ColumnMapping   `json:"saved_column_mapping,omitempty"`
	UnmappedColumnPolicy string            `json:"unmapped_column_policy,omitempty"` // warn | error | ignore
	CreatedAt            time.Time         `json:"created_at"`
	UpdatedAt            time.Time         `json:"updated_at"`
}

// AssetSink describes where ingested data is written.
type AssetSink struct {
	DatasourceAPIName APIName `json:"datasource"`
	Schema            string  `json:"schema,omitempty"`
	Table             string  `json:"table"`
}

// QualityRule is an asset-level data quality predicate.
type QualityRule struct {
	APIName     APIName             `json:"api_name"`
	Kind        QualityRuleKind     `json:"kind"`
	Property    APIName             `json:"property,omitempty"`
	Severity    QualityRuleSeverity `json:"severity"`
	Args        json.RawMessage     `json:"args,omitempty"`
	Description string              `json:"description,omitempty"`
	Metadata    map[string]any      `json:"metadata,omitempty"`
}

// QualityRuleKind selects the evaluator namespace. Lattice treats this as an
// open string so integrations can register domain-specific rules.
type QualityRuleKind string

const (
	QualityRuleKindNotNull       QualityRuleKind = "not_null"
	QualityRuleKindUnique        QualityRuleKind = "unique"
	QualityRuleKindRange         QualityRuleKind = "range"
	QualityRuleKindAllowedValues QualityRuleKind = "allowed_values"
	QualityRuleKindRegex         QualityRuleKind = "regex"
	QualityRuleKindCustomCEL     QualityRuleKind = "custom_cel"
)

// QualityRuleSeverity decides whether failures block the load.
type QualityRuleSeverity string

const (
	QualityRuleSeverityError   QualityRuleSeverity = "error"
	QualityRuleSeverityWarning QualityRuleSeverity = "warning"
)

// ColumnMapping maps a source file column to a target asset property.
type ColumnMapping struct {
	SourceColumn   string              `json:"source_column"`
	TargetProperty APIName             `json:"target_property"`
	Required       bool                `json:"required,omitempty"`
	TypeCoercion   string              `json:"type_coercion,omitempty"`
	ValueMapping   map[string]string   `json:"value_mapping,omitempty"`
	Transforms     []PropertyTransform `json:"transforms,omitempty"`
}

// AssetDependency captures a declared relationship to another asset or data
// producer so callers can build dependency graphs and impact analysis.
type AssetDependency struct {
	Kind        string         `json:"kind"`
	Target      string         `json:"target"`
	Description string         `json:"description,omitempty"`
	Metadata    map[string]any `json:"metadata,omitempty"`
}

File: crates/tesela-ir/src/asset.rs
//! Asset, column mapping, lineage, and environment types.

use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;
use tesela_core::{ApiName, Value};

use crate::Property;

/// A dataset asset produced by the upload pipeline.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Asset {
    /// API name.
    pub api_name: ApiName,
    /// Display name.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub display: Option<String>,
    /// Description.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub description: Option<String>,
    /// Sink destination.
    pub sink: AssetSink,
    /// Properties (schema of the asset).
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    pub properties: Vec<Property>,
    /// Tags.
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    pub tags: Vec<String>,
    /// Deprecation timestamp.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub deprecated_at: Option<String>,
    /// Metadata.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub metadata: Option<BTreeMap<String, Value>>,
    /// Lineage edges.
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    pub lineage: Vec<LineageEdge>,
}

/// Asset sink configuration.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AssetSink {
    /// Target datasource.
    pub datasource: ApiName,
    /// Target resource (table, bucket, etc.).
    pub resource: String,
}

/// Column mapping for upload/ingestion.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ColumnMapping {
    /// Source column name.
    pub source_column: String,
    /// Target property API name.
    pub target_property: ApiName,
    /// Whether this mapping is required.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub required: Option<bool>,
    /// Type coercion hint.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub type_coercion: Option<String>,
    /// Value mappings (source value -> target value).
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub value_mapping: Option<BTreeMap<String, String>>,
}

/// A lineage edge.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct LineageEdge {
    /// Source entity.
    pub source: ApiName,
    /// Kind of relationship (produces, consumes, derives_from).
    pub kind: String,
    /// Metadata.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub metadata: Option<BTreeMap<String, Value>>,
}

/// An environment configuration (dev, staging, prod).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Environment {
    /// API name.
    pub api_name: ApiName,
    /// Display name.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub display: Option<String>,
    /// Configuration values.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub config: Option<BTreeMap<String, Value>>,
    /// Metadata.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub metadata: Option<BTreeMap<String, Value>>,
}
